File: others/appendix/list/src/ListEx.hs
-- ListEx.hs
-- Copyright (C) 2012 Liu Xinyu (liuxinyu95@gmail.com)
--
-- This program is free software: you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation, either version 3 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program.  If not, see <http://www.gnu.org/licenses/>.

module ListEx where

import Data.List
import Test.QuickCheck -- for verification

atR :: [a] -> Int -> a
atR xs i = get xs (drop i xs) where
  get (x:_) [_] = x
  get (_:xs) (_:ys) = get xs ys

-- lastAt with zip
lastAt k xs = (fst . last) $ zip xs (drop k xs)

insertAt :: [a] -> Int -> a -> [a]
insertAt xs 0 y = y:xs
insertAt [] i y = [y]
insertAt (x:xs) i y = x : insertAt xs (i-1) y

permutation xs n r=  if length xs  <=  n - r
        then [[]]
        else [x:ys | x  <-  xs, ys  <-  permutation (delete x xs) n r]

-- Given a list of n elements, pick r from it to for permutation.
perm xs r | r == 0 || length xs < r = [[]]
          | otherwise = [ x:ys | x <-xs, ys <- perm (delete x xs) (r-1)]

-- Given a list of n elements, pick r from it to form combination.
comb _  0 = []
comb xs 1 = [[x] | x <- xs]
comb xs@(x:xs') r | length xs < r = []
                  | otherwise = [x:ys | ys <- comb xs' (r-1)] ++ comb xs' r

-- speakInt n = map speak (reverse $ zip ds ["", "thousand", "million", "billion"]) where
--   ds = reverse (digits n)
--   words = ["", "one", "two", "three", "four"]
--   speak (n, unit) | n < 20

digits = dec [] where
  dec ds n | n == 0 = ds
           | otherwise = dec ((n `mod` 1000) : ds) (n `div` 1000)

safeTake n [] = []
safeTake n (x:xs) | n <= 0 = []
                  | otherwise = x : safeTake (n - 1) xs

safeDrop n [] = []
safeDrop n (x:xs) | n <= 0 = x:xs
                  | otherwise = drop (n - 1) xs

sublist from cnt = take cnt . drop (from - 1)

slice from to = drop (from - 1) . take to

suffixes [] = [[]]
suffixes xs@(_:ys) =  xs : suffixes ys

sufs = foldr f [[]] where
  f x xss@(xs:_) = (x:xs) : xss

concats [] = []
concats ([]:xss) = concat xss
concats ((x:xs):xss) = x : concat (xs:xss)

foldr2 f z [] ys = z
foldr2 f z xs [] = z
foldr2 f z (x:xs) (y:ys) = foldr2 f (f x y z) xs ys

zip2 = foldr2 f [] where
  f x y xys = (x, y) : xys

iota1 = iota2 1

iota2 m n = iota' [] n where
  iota' ns n | n < m = ns
             | otherwise = iota' (n : ns) (n - 1)

iota m n a | m <= n = m : iota (m + a) n a
           | otherwise = []

iota4 m = m : iota4 (m + 1)

dedup  = foldr f [] where
  f x xs = x : filter (x /=) xs

-- test

prop_rindex :: [Int] -> Bool
prop_rindex xs = xs == (map (atR xs) $ reverse [0..length xs -1])

prop_insertAt :: [Int] -> Int -> Int -> Property
prop_insertAt xs i x = (0 <= i) ==> (insertAt xs i x) == (let (as, bs) = splitAt i xs in as ++ x:bs)

prop_concat :: [[Int]] -> Bool
prop_concat xss = concat xss == concats xss

prop_tails :: [Int] -> Bool
prop_tails xs = suffixes xs == xss && sufs xs == xss where
  xss = tails xs

prop_iota1 :: Int -> Bool
prop_iota1 z = let n = abs z in iota1 n == (take n $ iterate (+1) 1)

prop_iota2 :: Int -> Int -> Bool
prop_iota2 x y = iota2 m n == (take (n - m + 1) $ iterate (+1) m) where
  m = min (abs x) (abs y)
  n = max (abs x) (abs y)

prop_iota3 :: Int -> Int -> Bool
prop_iota3 x y = all (\a -> iota m n a == (takeWhile (<= n) $ iterate (+a) m)) [1..(n - m)]
  where
    m = min (abs x) (abs y)
    n = max (abs x) (abs y)

prop_iota4 :: Int -> Bool
prop_iota4 z = let m = abs z in take 10 (iota4 m) == (take 10 $ iterate (+1) m)

prop_nub :: [Int] -> Bool
prop_nub xs = nub xs == dedup xs

File: others/appendix/list/src/listex.py
#!/usr/bin/python

# list-imp.py
# Copyright (C) 2012 Liu Xinyu (liuxinyu95@gmail.com)
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# Imperative list manipulations
import random

class List:
    def __init__(self, x = None, xs = None):
        self.key = x
        self.next = xs

def cons(x, xs):
    return List(x, xs)

def length(xs):
    n = 0
    while xs is not None:
        n = n + 1
        xs = xs.next
    return n

def append(xs, x):
    xs = cons(None, xs)
    r = xs
    while xs.next is not None:
        xs = xs.next
    xs.next = List(x)
    return r.next

def takeM(n, xs):
    d = len(xs) - n
    while d > 0 and xs:
        xs.pop()
        d = d - 1
    return xs

def dropM(n, xs):
    if n <= 0:
        return xs
    for i in range(n, len(xs)):  # shift all elements ahead n
        xs[i - n] = xs[i]
    while n > 0 and xs:
        xs.pop()
        n = n - 1
    return xs

def mapL(f, xs):
    ys = prev = List()
    while xs is not None:
        prev.next = List(f(xs.key))
        prev = prev.next
        xs = xs.next
    return ys.next

def span(p, xs):
    ys = xs
    last = None
    while xs is not None and p(xs.key):
        last = xs
        xs = xs.next
    if last is None:
        return (None, xs)
    last.next = None
    return (ys, xs)

def group1(p, xs):
    if xs is None:
        return List(None)
    (x, xs) = (xs.key, xs.next)
    g = List(x)
    G = List(g)
    while xs is not None:
        y = xs.key
        if p(x, y):
            g = append(g, y)
        else:
            g = List(y)
            G = append(G, g)
        x = y
        xs = xs.next
    return G

def find(p, xs):
    while xs is not None:
        if p(xs.key):
            return xs
        xs = xs.next
    return None

def filterL(p, xs):
    ys = None
    while xs:
        if p(xs.key):
            ys = cons(xs.key, ys)
        xs = xs.next
    return reverse(ys)

def reverse(xs):
    ys = None
    while xs:
        ys = cons(xs.key, ys)
        xs = xs.next
    return ys

def lookupAssoc(x, assoc):
    for k, v in assoc:
        if x == k:
            return v
    return None

def zipL(xs, ys):
    zs = None
    while xs and ys:
        zs = cons((xs.key, ys.key), zs)
        xs = xs.next
        ys = ys.next
    return reverse(zs)

def fromList(xs):
    ys = None
    for x in reversed(xs):
        ys = cons(x, ys)
    return ys

def toList(xs):
    ys = []
    while xs is not None:
        ys.append(xs.key)
        xs = xs.next
    return ys

# testing
def randlist(n):
    return random.sample(range(n), random.randint(0, n))

def test_zip():
    for _ in range(100):
        xs = randlist(100)
        ys = randlist(100)
        assert toList(zipL(fromList(xs), fromList(ys))) == list(zip(xs, ys))
    print("tested zip")

def test_filter():
    even = lambda x: x % 2 == 0
    for _ in range(100):
        lst = randlist(100)
        assert toList(filterL(even, fromList(lst))) == list(filter(even, lst))
    print("tested filter")

def test_map():
    negate = lambda x: -x
    inc = lambda x: 1 + x
    for _ in range(100):
        lst = randlist(100)
        assert all([toList(mapL(f, fromList(lst))) == list(map(f, lst)) for f in [negate, inc, abs]])
    print("tested map")

def test_span():
    for _ in range(100):
        lst = list(range(100))
        random.shuffle(lst)
        (xs, ys) = span(lambda x: x < 50, fromList(lst))
        if xs is not None:
            assert any(toList(mapL(lambda x: x < 50, xs)))
        if ys is not None:
            assert not ys.key < 50
        assert length(xs) + length(ys) == len(lst)
    print("tested span")

def test_find():
    for _ in range(100):
        lst = randlist(100)
        x = random.randint(0, 100)
        if x in lst:
            assert x == find(lambda a : x == a, fromList(lst)).key
        else:
            assert None == find(lambda a : x == a, fromList(lst))
    print("tested find")

def test_group1():
    gs = group1(lambda x, y: x == y, fromList("mississipi"))
    while gs is not None:
        print(toList(gs.key), end = '')
        gs = gs.next
    print("tested groupping")

def test_sublist():
    for _ in range(100):
        xs = randlist(100)
        n = random.randint(-10, 110)
        assert takeM(n, xs) == xs[:n]
        assert dropM(n, xs) == xs[n:]
    print("tested sublist")

if __name__ == "__main__":
    test_map()
    test_span()
    test_group1()
    test_find()
    test_sublist()
    test_filter()
    test_zip()
